routing/router.go

```diff
@@ -1620,131 +1665,208 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
 	// critical error during path finding.
 	var lastError error
 	for {
-		// Before we attempt this next payment, we'll check to see if
-		// either we've gone past the payment attempt timeout, or the
-		// router is exiting. In either case, we'll stop this payment
-		// attempt short.
-		select {
-		case <-timeoutChan:
-			errStr := fmt.Sprintf("payment attempt not completed "+
-				"before timeout of %v", payAttemptTimeout)
 
-			return [32]byte{}, nil, newErr(
-				ErrPaymentAttemptTimeout, errStr,
-			)
+		// If this payment had no existing payment ID, we make a new attempt.
+		if attempt == nil {
+			// Before we attempt this next payment, we'll check to see if
+			// either we've gone past the payment attempt timeout, or the
+			// router is exiting. In either case, we'll stop this payment
+			// attempt short.
+			select {
+			case <-timeoutChan:
```

Comment:
AFAICT, this is the only place that the timeout channel is checked. Therefore, we should move it outside of this if statement to ensure that it's always checked just as it was in the prior diff at the top of the loop.

Reply:
This block can also likely be swept into its own method.

Reply:
If `attempt != nil` we cannot timeout the payment, as it might already be forwarded onto the network.

Reply:
+1 on extracting methods. `sendPayment` is now around 350 lines. I would like to re-suggest to move things into payment session, to prevent having many arguments for extracted methods.

Reply:
Moved into struct `paymentLifecycle`.